### PR TITLE
Port TaskObserver into PTVS

### PR DESCRIPTION
**Pull request by AlexanderSher**
This approach helps tests catch exceptions in tasks that aren't awaited or run too for long.

For example, imagine the following code:
```cs
public void Run() {
    RunThenFailAsync().DoNotWait();
    RunForeverAsync().DoNotWait();
}

private async Task RunThenFailAsync() {
    await Task.Delay(1000);
    throw new InvalidOperationException();
}

private async Task RunForeverAsync() {
    while (true) {
        await Task.Delay(500);
    }
}
```

If we create a simple test for it, it will successfully pass:
```cs
[TestMethod]
public void SomeTypeTest() {
    new SomeType().Run();
}
```
image

But, with `TaskObserver` enabled:
```cs
[TestInitialize]
public void Initialize() => MSTestEnvironment.TestInitialize(5);

[TestCleanup]
public void Cleanup() => MSTestEnvironment.TestCleanup();

[TestMethod]
public void SomeTypeTest() {
    new SomeType().Run();
}
```
we will get both tasks in the output:
```
Test Failed - SomeTypeTest
Message: One task that has been started during test run is still not completed:

   at void SomeType.Run() in C:\Temp\SomeType.cs:line 29
   at void Tests.SomeTypeTest() in C:\Temp\Tests.cs:line 44
   at object RuntimeMethodHandle.InvokeMethod(object target, object[] arguments, Signature sig, bool constructor)
   at object System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(object obj, object[] parameters, object[] arguments)
   at object System.Reflection.RuntimeMethodInfo.Invoke(object obj, BindingFlags invokeAttr, Binder binder, object[] parameters, CultureInfo culture)

Also, one not awaited tasks has failed:

InvalidOperationException: Operation is not valid due to the current state of the object.
   at async Task SomeType.RunThenFailAsync() in C:\Temp\SomeType.cs:line 34
   --- Task stack trace: ---
   at void SomeType.Run() in C:\Temp\SomeType.cs:line 28
   at void Tests.SomeTypeTest() in C:\Temp\Tests.cs:line 44
   at object RuntimeMethodHandle.InvokeMethod(object target, object[] arguments, Signature sig, bool constructor)
   at object System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(object obj, object[] parameters, object[] arguments)
   at object System.Reflection.RuntimeMethodInfo.Invoke(object obj, BindingFlags invokeAttr, Binder binder, object[] parameters, CultureInfo culture)
```

### CODE REVIEW

File: Common/Tests/Utilities/MSTestEnvironment.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Common = Microsoft.PythonTools.Infrastructure;
+using Analysis = Microsoft.PythonTools.Analysis.Infrastructure;
```
Comment (zooba) [inline Common/Tests/Utilities/MSTestEnvironment.cs]: This should go in `TestUtilities.Python.Analysis` in order to take dependencies on PythonTools assemblies. (Or we should move the entire TestUtilities project out of the Common directory - I'm 99% sure nobody else is using this project now, but at one point a couple of external companies were using it.)

File: Python/Tests/Core/AssemblySetup.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// permissions and limitations under the License.
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using TestUtilities;
+
+namespace PythonToolsTests {
+    [TestClass]
+    public sealed class AssemblySetup {
+        [AssemblyInitialize]
+        public static void Initialize(TestContext testContext) => MSTestEnvironment.Initialize();
```
Comment (zooba) [inline Python/Tests/Core/AssemblySetup.cs]: We'll also need something similar in the test runner interop module for the UI tests.

File: Common/Tests/Utilities/MSTestEnvironment.cs (hunk 1 above)
Comment (AlexanderSher) [reply Common/Tests/Utilities/MSTestEnvironment.cs]: It is dependent only by projects in `Test` folder:

image

Do you want me to move it?

File: Python/Tests/Core/AssemblySetup.cs (hunk 2 above)
Comment (AlexanderSher) [reply Python/Tests/Core/AssemblySetup.cs]: I don't get it. `TestUtilities.UI` references `TestUtilities` directly, can't we just call `MSTestEnvironment.Initialize()` in the `HostedPythonToolsTestRunner` constructor?

File: Common/Tests/Utilities/MSTestEnvironment.cs (hunk 1 above)
Comment (zooba) [reply Common/Tests/Utilities/MSTestEnvironment.cs]: Within our repo, this is true, but there are other projects besides PTVS that use this. If we move it, they have to stop using it, whereas if we just move the new additions into the Python-specific project then it doesn't affect the Common one.

File: Python/Tests/Core/AssemblySetup.cs (hunk 2 above)
Comment (zooba) [reply Python/Tests/Core/AssemblySetup.cs]: Yes. I don't see that change in this PR.

Comment (zooba) [reply Python/Tests/Core/AssemblySetup.cs]: But it needs to be called on the VS side, not the MSTest side.

File: Common/Tests/Utilities/MSTestEnvironment.cs (hunk 1 above)
Comment (AlexanderSher) [reply Common/Tests/Utilities/MSTestEnvironment.cs]: Ok. So `TestUtilities.Python.Analysis` or `TestUtilities.Python`?

File: Python/Tests/Core/AssemblySetup.cs (hunk 2 above)
Comment (AlexanderSher) [reply Python/Tests/Core/AssemblySetup.cs]: Ok

File: Common/Tests/Utilities/MSTestEnvironment.cs (hunk 1 above)
Comment (zooba) [reply Common/Tests/Utilities/MSTestEnvironment.cs]: TestUtilities.Python.Analysis (which is the "non-UI" test project... the naming is not great :) )

**Comment by AlexanderSher** [comment]
Ping

File: Python/Tests/Utilities.Python.Analysis/MSTestEnvironment.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using Common = Microsoft.PythonTools.Infrastructure;
+using Analysis = Microsoft.PythonTools.Analysis.Infrastructure;
+
+namespace TestUtilities.Python {
+    public sealed class MSTestEnvironment : TestEnvironment, Common.ITestEnvironment, Analysis.ITestEnvironment {
+        public static void Initialize() {
+            var isntance = new MSTestEnvironment();
```
Comment (zooba) [inline Python/Tests/Utilities.Python.Analysis/MSTestEnvironment.cs]: *instance

**Pull request merged**